### PR TITLE
changeMasterSecretPassphrase is now being called in AsyncTask

### LINKED ISSUE #2768: PassphraseChangeActivity not using AsyncTask for MasterSecretUtil.changeMasterSecretPassphrase()
`PassphraseChangeActivity` is calling `MasterSecretUtil.changeMasterSecretPassphrase()` on the UI thread.

https://github.com/WhisperSystems/TextSecure/blob/master/src/org/thoughtcrime/securesms/PassphraseChangeActivity.java#L100

### PR DESCRIPTION
**Pull request by adrianchifor**
Created new ChangeMasterSecretPassphrase AsyncTask class to handle the changeMasterSecretPassphrase method outside the UI thread. 

Also moved the try-catch block for changeMasterSecretPassphrase from verifyAndSavePassphrases method to AsyncTask class so we don't have the other stuff that don't throw any exceptions in the try block.

Fixes #2768

### CODE REVIEW

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 1]
```diff
@@ -124,6 +117,52 @@ public void onClick(View v) {
     }
   }
 
+  private class ChangeMasterSecretPassphrase extends AsyncTask<Void, Void, Void> {
+    private MasterSecret masterSecret;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: final

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 2]
```diff
@@ -124,6 +117,52 @@ public void onClick(View v) {
...
+    private InvalidPassphraseException exception;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: this is pretty gross, don't use an instance variable as a rendevouz

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 3]
```diff
@@ -124,6 +117,52 @@ public void onClick(View v) {
...
+    private EditText originalPassphrase;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: why is this an instance variable?

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 4]
```diff
@@ -124,6 +117,52 @@ public void onClick(View v) {
...
+    private Context context;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: final

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 5]
```diff
@@ -124,6 +117,52 @@ public void onClick(View v) {
...
+    private String original, passphrase;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: separate lines

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 6]
```diff
@@ -124,6 +117,52 @@ public void onClick(View v) {
...
+
+    public ChangeMasterSecretPassphrase(Context context, String original, String passphrase) {
+      this.context = context;
+      this.original = original;
+      this.passphrase = passphrase;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: align

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 7]
```diff
@@ -124,6 +117,52 @@ public void onClick(View v) {
...
+    }
+
+    @Override
+    protected Void doInBackground(Void... params) {
+      try {
+        this.masterSecret = MasterSecretUtil.changeMasterSecretPassphrase(context, original, passphrase);
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: return this, don't use an instance variable

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 8]
```diff
@@ -124,6 +117,52 @@ public void onClick(View v) {
...
+      }
+
+      return null;
+    }
+
+    @Override
+    protected void onPostExecute(Void aVoid) {
+      this.originalPassphrase = (EditText) PassphraseChangeActivity.this.findViewById(R.id.old_passphrase);
+
+      // If exception not thrown and master secret not null -> setMasterSecret
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: no comments, ever

**Comment by adrianchifor** [comment]
@moxie0 Oups, my bad. Thanks for the suggestions! Fixed the code. Did a quick test and works fine.

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 9]
```diff
@@ -124,6 +117,44 @@ public void onClick(View v) {
     }
   }
 
+  private class ChangeMasterSecretPassphrase extends AsyncTask<Void, Void, MasterSecret> {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: ChangePassphraseTask

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 10]
```diff
@@ -124,6 +117,44 @@ public void onClick(View v) {
...
+      }
+    }
+
+    @Override
+    protected void onPostExecute(MasterSecret masterSecret) {
+      if (masterSecret != null) {
+        setMasterSecret(masterSecret);
+      }
+
+      else {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: one line

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 11]
```diff
@@ -124,6 +117,44 @@ public void onClick(View v) {
...
+    @Override
+    protected MasterSecret doInBackground(Void... params) {
+      try {
+        MasterSecret masterSecret = MasterSecretUtil.changeMasterSecretPassphrase(context, original, passphrase);
+        TextSecurePreferences.setPasswordDisabled(context, false);
+
+        return masterSecret;
+
+      } catch (InvalidPassphraseException e) {
+        return null;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: log the exception

**Comment by adrianchifor** [comment]
@moxie0 Thanks again! Updated the code. App builds fine.

**Comment by moxie0** [comment]
it looks like this creates a scenario where the user might double-click the button and kick off two async tasks

**Comment by adrianchifor** [comment]
Hmm, that's a good point. I'll fix that tomorrow.

**Comment by adrianchifor** [comment]
@moxie0 'Ok' button is now being disabled on pre-execute and re-enabled on post-execute.

File: src/org/thoughtcrime/securesms/PassphraseChangeActivity.java [hunk 12]
```diff
@@ -124,6 +118,50 @@ public void onClick(View v) {
     }
   }
 
+  private class ChangePassphraseTask extends AsyncTask<Void, Void, MasterSecret> {
+    private final Context context;
+    private String original;
+    private String passphrase;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseChangeActivity.java]: original and passphrase should be parameters rather than instance vars

**Comment by adrianchifor** [comment]
@moxie0 Updated

**Pull request closed without merging**